idaes/core/solvers/petsc.py

```diff
@@ -519,6 +520,30 @@ def petsc_dae_by_time_element(
     return res_list
 
 
+def calculate_time_derivatives(m, time):
+    """Calculate the derivative values from the discretization equations.
+
+    Args:
+        m (Block): Pyomo model block
+        time (ContinuousSet): Time set
+
+    Returns:
+        None
+    """
+    for var in m.component_objects(pyo.Var):
+        if isinstance(var, pyodae.DerivativeVar):
+            if time in ComponentSet(var.get_continuousset_list()):
+                parent = var.parent_block()
+                name = var.local_name + "_disc_eq"
+                disc_eq = getattr(parent, name)
+                for i, v in var.items():
+                    try:
+                        if disc_eq[i].active:
+                            calculate_variable_from_constraint(v, disc_eq[i])
+                    except KeyError:
+                        pass # discretization equation may not exist at first time
```

Comment:
Should we check that this is the first time point in this case, and raise an error if it is not?

Reply:
That should be covered by the KeyError exception.  The first time point shouldn't have a discretization equation.

Reply:
Oh yeah, now I understand.  I'm not sure I have a clear way to know which ones should be missing with all the options.  I guess the ones that are missing should be missing.  The DAE transformation writes them so there probably not a lot of room for users to just do something weird.

Reply:
You could check this without try/except by just checking `i in disc_eq`